consul-lambda-registrator/extension/extension.go

```diff
@@ -0,0 +1,312 @@
+	"github.com/hashicorp/terraform-aws-consul-lambda-registrator/consul-lambda-registrator/structs"
+	"github.com/hashicorp/terraform-aws-consul-lambda-registrator/consul-lambda-registrator/trace"
+)
+
+type Config struct {
+	MeshGatewayURI      string        `envconfig:"MESH_GATEWAY_URI" required:"true"`
+	ExtensionDataPrefix string        `envconfig:"EXTENSION_DATA_PREFIX" required:"true"`
+	Datacenter          string        `envconfig:"DATACENTER"`
+	ServiceName         string        `envconfig:"SERVICE_NAME"`
+	ServiceNamespace    string        `envconfig:"SERVICE_NAMESPACE"`
+	ServicePartition    string        `envconfig:"SERVICE_PARTITION"`
+	ServiceUpstreams    []string      `envconfig:"SERVICE_UPSTREAMS"`
+	RefreshFrequency    time.Duration `envconfig:"REFRESH_FREQUENCY" default:"5m"`
+
+	Store  ParamGetter
+	Events EventProcessor
+	Logger hclog.Logger
+}
+
+type ParamGetter interface {
+	// Get the value for the given key.
+	Get(ctx context.Context, key string) (string, error)
+}
+
+type EventProcessor interface {
+	// Register the event processor.
+	Register(ctx context.Context, i interface{}) error
+	// ProcessEvents handles events until the provided context is cancelled or an error occurs.
+	ProcessEvents(ctx context.Context) error
+}
+
+type Extension struct {
+	*Config
+	service    structs.Service
+	proxy      *proxy.Server
+	proxyMutex sync.Mutex
+	data       structs.ExtensionData
+	upstreams  []structs.Service
+}
+
+// NewExtension returns an instance of the Extension from the given configuration.
+func NewExtension(cfg *Config) *Extension {
+	ext := &Extension{
+		Config: cfg,
+		service: structs.Service{
+			Datacenter: cfg.Datacenter,
+			Name:       cfg.ServiceName,
+			Namespace:  cfg.ServiceNamespace,
+			Partition:  cfg.ServicePartition,
+		},
+	}
+	return ext
+}
+
+// Serve executes the main processing loop for the extension.
+// It initializes and starts the proxy server and starts monitoring for incoming
+// events from the Lambda runtime.
+// It periodically retrieves the extension data from the parameter store and updates
+// the proxy configuration for the configured upstreams as necessary.
+func (ext *Extension) Serve(ctx context.Context) error {
+	trace.Enter()
+	defer trace.Exit()
+
+	ctx, cancel := context.WithCancel(ctx)
+
+	// Cleanup on return. Cancel the context and close the proxy.
+	defer cancel()
+	defer func() {
+		ext.proxyMutex.Lock()
+		defer ext.proxyMutex.Unlock()
+		ext.closeProxy()
+	}()
+
+	proxyErrChan := make(chan error)
+	go ext.runProxy(ctx, proxyErrChan)
+
+	eventErrChan := make(chan error)
+	go ext.runEvents(ctx, eventErrChan)
+
+	// Block until either the proxy returns or until the event processing loop returns.
+	select {
+	case err := <-proxyErrChan:
+		return err
+	case err := <-eventErrChan:
+		return err
+	}
+}
+
+func (ext *Extension) runProxy(ctx context.Context, errChan chan error) {
+	refresh := time.NewTicker(ext.RefreshFrequency)
```

Comment:
Do you know how this works with Lambda freezes/unfreezes?

Reply:
Great question! I don't know how the ticker would work for freeze/unfreeze. I'll look into this. It may be better to use absolute time to tigger the refresh.

Reply:
I did some manual testing on this and it looks to me like it is working as expected.

I set the `REFRESH_FREQUENCY` to 10 seconds and invoked the function a bunch of times at random intervals between 1s and 10m. When the function is invoked every second the extension triggers the refresh at the configured interval. When the gap between invocations grows to the point that the runtime gets frozen (several seconds) the extension data and proxy refresh is triggered immediately once the function is invoked. And the refresh is only triggered once, meaning that the timer ticks don't accumulate.

Reply:
Yay, this is awesome. Thanks for verifying this.